flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceSplitEnumerator.java

```diff
@@ -0,0 +1,406 @@
+                subtaskId,
+                currentSourceIndex,
+                pendingSplits);
+        // TODO: test coverage for on demand split assignment
+        assignPendingSplits(subtaskId);
+        currentEnumerator.handleSplitRequest(subtaskId, requesterHostname);
+    }
+
+    @Override
+    public void addSplitsBack(List<HybridSourceSplit<SplitT>> splits, int subtaskId) {
+        LOG.debug("Adding splits back for subtask={} {}", subtaskId, splits);
+        // Splits returned can belong to multiple sources, after switching since last checkpoint
+        TreeMap<Integer, List<HybridSourceSplit<SplitT>>> splitsBySourceIndex = new TreeMap<>();
+
+        for (HybridSourceSplit<SplitT> split : splits) {
+            splitsBySourceIndex
+                    .computeIfAbsent(split.sourceIndex(), k -> new ArrayList<>())
+                    .add(split);
+        }
+
+        splitsBySourceIndex.forEach(
+                (k, splitsPerSource) -> {
+                    if (k == currentSourceIndex) {
+                        currentEnumerator.addSplitsBack(
+                                HybridSourceReader.unwrapSplits(splitsPerSource), subtaskId);
+                    } else {
+                        pendingSplits
+                                .computeIfAbsent(subtaskId, sourceIndex -> new TreeMap<>())
+                                .put(k, splitsPerSource);
+                        if (context.registeredReaders().containsKey(subtaskId)) {
+                            assignPendingSplits(subtaskId);
+                        }
+                    }
+                });
+    }
+
+    @Override
+    public void addReader(int subtaskId) {
+        LOG.debug("addReader subtaskId={}", subtaskId);
+        if (pendingSplits.isEmpty()) {
+            context.sendEventToSourceReader(subtaskId, new SwitchSourceEvent(currentSourceIndex));
+            LOG.debug("Adding reader {} to enumerator {}", subtaskId, currentSourceIndex);
+            currentEnumerator.addReader(subtaskId);
+        } else {
+            // Defer adding reader to the current enumerator until splits belonging to earlier
+            // enumerators that were added back have been processed
+            pendingReaders.add(subtaskId);
+            assignPendingSplits(subtaskId);
+        }
+    }
+
+    private void assignPendingSplits(int subtaskId) {
+        TreeMap<Integer, List<HybridSourceSplit<SplitT>>> splitsBySource =
+                pendingSplits.get(subtaskId);
+        if (splitsBySource != null) {
+            int sourceIndex = splitsBySource.firstKey();
+            List<HybridSourceSplit<SplitT>> splits =
+                    Preconditions.checkNotNull(splitsBySource.get(sourceIndex));
+            Preconditions.checkState(!splits.isEmpty());
+            LOG.debug("Assigning pending splits subtask={} {}", subtaskId, splits);
+            context.sendEventToSourceReader(subtaskId, new SwitchSourceEvent(sourceIndex));
```

Comment:
we send the switch event before `context.assignSplits`. that means at TM side, we can have two readers (from two sources) active at the same time, right? Just want to make sure I understand the behavior.

Reply:
There is only one reader active at any point in time. Because as part of recovery non-checkpointed splits are added back to the enumerator, those need to be processed by the respective source reader. So the underlying current enumerator is effectively suspended and HybridSourceSplitEnumerator replays the splits with the respective readers in the original order. I tried to capture that in the javadoc but probably need to expand it.

Reply:
assume it's an embarrassingly parallel DAG with region recovery enabled. Let's see this scenario below can happen or not.
- T1: checkpoint N completed
- T2: switched from Iceberg to Kafka source
- T3: subtask/reader X failed and restarted. Iceberg splits are added back, subtask X requested and got Iceberg splits. This happened in the same time while other subtasks are reading from Kafka. That means the hybrid source can read from Kafka and Iceberg simultaneously during failure recovery scenario.

Reply:
also how does `handleSplitRequest` work in the above scenario. 
- it will first assign all recovered/added-back Iceberg splits (which is correct)
- but then the next step is to forward the request to the current enumerator, which is the Kafka source. is this correct?

```
        assignPendingSplits(subtaskId);
        currentEnumerator.handleSplitRequest(subtaskId, requesterHostname);
```

Reply:
This path isn't exercised in the ITCase and I don't think this is correct. The `TODO` (test coverage) needs to be addressed.

When `assignPendingSplits` sent splits to the subtask then the call should probably return and not delegate to the underlying enumerator, as that could result in splits being sent to a previous reader.

Reply:
Fixed and covered with unit test.